### PR TITLE
Require separation of API and SDK artifacts

### LINKED ISSUE #4084: Require separation of API and SDK artifacts
**What are you trying to achieve?**

The spec [states today](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/library-guidelines.md#requirements):

> 1. The OpenTelemetry API must be well-defined and clearly decoupled from the implementation. This allows end users to consume API only without also consuming the implementation (see points 2 and 3 for why it is important).

> 2. Third party libraries and frameworks that add instrumentation to their code will have a dependency only on the API of OpenTelemetry client. The developers of third party libraries and frameworks do not care (and cannot know) what specific implementation of OpenTelemetry is used in the final application.

The spirit of the 2nd clause has always been to not require any additional dependencies when taking a dependency on the API library. However, some implementations (e.g. https://github.com/open-telemetry/opentelemetry-swift/issues/486) release both API and SDK as a single artifact that makes it impossible for the user to not pull all the SDK dependencies as well.

I would like to propose the clarification to the spec which makes zero-dependency requirement explicit, by adding this text to the 2nd point:

> To enable this clean separation the API and SDK libraries MUST be provided as independent artifacts.

**Open questions**

Are there other SDKs besides Swift that bundle API and SDK into a single artifact?

### PR DESCRIPTION
**Pull request by yurishkuro**
Resolves #4084.

The Technical Committee discussed this issue on 2024-07-10 and recommends to make this change, because it aligns with the fundamental project philosophy and we want to make sure as many language SIGs as possible meet this requirement (and almost all of them already do). 

This change could make some SIGs non-compliant. They should clearly document this non-compliance and the reasoning. The TC strongly recommends that the non-compliant SIGs find ways to achieve artifact separation. In case it is completely impossible due to limitations of a specific ecosystem / package managers, this could become a permanent (documented) exception.

### CODE REVIEW

**Review by cijothomas — APPROVED** [review]

**Comment by jpkrohling** [comment]
Is there a reason behind this PR being closed?

**Comment by joaopgrassi** [comment]
Just to link things, this applies to CPP https://github.com/open-telemetry/opentelemetry-cpp/issues/2543

File: specification/library-guidelines.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ _Note to OpenTelemetry client Authors:_ OpenTelemetry specification, API and SDK
 
 1. The OpenTelemetry API must be well-defined and clearly decoupled from the implementation. This allows end users to consume API only without also consuming the implementation (see points 2 and 3 for why it is important).
 
-2. Third party libraries and frameworks that add instrumentation to their code will have a dependency only on the API of OpenTelemetry client. The developers of third party libraries and frameworks do not care (and cannot know) what specific implementation of OpenTelemetry is used in the final application.
+2. Third party libraries and frameworks that add instrumentation to their code will have a dependency only on the API of OpenTelemetry client. The developers of third party libraries and frameworks do not care (and cannot know) what specific implementation of OpenTelemetry is used in the final application. To enable this clean separation, the OpenTelemetry API and SDK libraries MUST be provided as independent artifacts.
```
Comment (pellared) [inline specification/library-guidelines.md:17]: I suggest to use the "OpenTelemetry client" term consistently.

```suggestion
2. Third party libraries and frameworks that add instrumentation to their code will have a dependency only on the API of OpenTelemetry client. The developers of third party libraries and frameworks do not care (and cannot know) what specific implementation of OpenTelemetry is used in the final application. To enable this clean separation, the OpenTelemetry API and SDK clients MUST be provided as independent artifacts.
```

**Review by pellared — APPROVED** [review]

File: specification/library-guidelines.md (hunk 1 above)
Comment (yurishkuro) [reply specification/library-guidelines.md:17]: I think library is much more accurate in this context. "client" is not even mentioned in the glossary, I don't know when we started using that, and "client" refers to functionality, not to the packaging artifact. E.g. the Telemetry SDK in the glossary is referred to as a "library": https://github.com/open-telemetry/opentelemetry-specification/blob/3efed53cc1842b2a1d74e69c468e268649cf1d34/specification/glossary.md?plain=1#L115-L117

Comment (yurishkuro) [reply specification/library-guidelines.md:17]: I am not changing existing wording. In fact, the existing wording is what causes the confusion in the first place, when it says "will have a dependency only on the API of OpenTelemetry client" (which sounds like "client" is a single artifact).

**Review by tsloughter — APPROVED** [review]

**Review by reyang — APPROVED** [review]

**Review by jmacd — APPROVED** [review]

**Review by arminru — APPROVED** [review]

**Pull request merged**